src/app/clusters/commodity-price-server/commodity-price-server.cpp

```diff
@@ -58,21 +64,48 @@ bool Instance::HasFeature(Feature aFeature) const
 // AttributeAccessInterface
 CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
 {
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    DataModel::Nullable<Structs::CommodityPriceStruct::Type> priceStruct;
+    chip::Platform::ScopedMemoryBuffer<Structs::CommodityPriceStruct::Type> forecastBuffer;
+    DataModel::List<const Structs::CommodityPriceStruct::Type> forecastList;
+
     switch (aPath.mAttributeId)
     {
     case TariffUnit::Id:
-        return aEncoder.Encode(mDelegate.GetTariffUnit());
+        return aEncoder.Encode(mTariffUnit);
     case Currency::Id:
-        return aEncoder.Encode(mDelegate.GetCurrency());
+        return aEncoder.Encode(mCurrency);
     case CurrentPrice::Id:
-        return aEncoder.Encode(mDelegate.GetCurrentPrice());
+        // Call GetDetailedPriceRequest with details = 0 to strip out .components and .description if present
+        ReturnErrorOnFailure(GetDetailedPriceRequest(0, priceStruct));
+        err = aEncoder.Encode(priceStruct);
+
+        return err;
     case PriceForecast::Id:
         /* FORE - Forecasting */
         if (!HasFeature(Feature::kForecasting))
         {
             return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
         }
-        return aEncoder.Encode(mDelegate.GetPriceForecast());
+
+        if (!forecastBuffer.Calloc(kMaxForecastEntries))
```

Comment:
Could also have the callee allocate this, at whatever size it really needs.  But either way.